tool/tbot/config/config.go

```diff
@@ -0,0 +1,254 @@
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package config
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	CONFIG_DEFAULT_CERTIFICATE_TTL = 60 * time.Minute
+	CONFIG_DEFAULT_RENEW_INTERVAL  = 20 * time.Minute
+)
+
+var log = logrus.WithFields(logrus.Fields{
+	trace.Component: teleport.ComponentTBot,
+})
+
+// CLIConf is configuration from the CLI.
+type CLIConf struct {
+	ConfigPath string
+
+	Debug      bool
+	AuthServer string
+
+	// DataDir stores the bot's internal data.
+	DataDir string
+
+	// DestinationDir stores the generated end-user certificates.
+	DestinationDir string
+
+	// CAPins is a list of pinned SKPI hashes of trusted auth server CAs, used
+	// only on first connect.
+	CAPins []string
+
+	// Token is a bot join token.
+	Token string
+
+	// RenewInterval is the interval at which certificates are renewed, as a
+	// time.ParseDuration() string. It must be less than the certificate TTL.
+	RenewInterval time.Duration
+
+	// CertificateTTL is the requested TTL of certificates. It should be some
+	// multiple of the renewal interval to allow for failed renewals.
+	CertificateTTL time.Duration
+}
+
+// OnboardingConfig contains values only required on first connect.
+type OnboardingConfig struct {
+	// Token is a bot join token.
+	Token string `yaml:"token"`
+
+	// CAPath is an optional path to a CA certificate.
+	CAPath string
+
+	// CAPins is a list of certificate authority pins, used to validate the
+	// connection to the Teleport auth server.
+	CAPins []string `yaml:"ca_pins"`
+}
+
+// BotConfig is the bot's root config object.
+type BotConfig struct {
+	Onboarding   *OnboardingConfig    `yaml:"onboarding,omitempty"`
+	Storage      StorageConfig        `yaml:"storage,omitempty"`
+	Destinations []*DestinationConfig `yaml:"destinations,omitempty"`
+
+	Debug          bool          `yaml:"debug"`
+	AuthServer     string        `yaml:"auth_server"`
+	CertificateTTL time.Duration `yaml:"certificate_ttl"`
+	RenewInterval  time.Duration `yaml:"renew_interval"`
+}
+
+func (conf *BotConfig) CheckAndSetDefaults() error {
+	if err := conf.Storage.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	for _, dest := range conf.Destinations {
+		if err := dest.CheckAndSetDefaults(); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	if conf.AuthServer == "" {
```

Comment:
nit: but I generally prefer to do these checks first, so that we avoid mutations via any of the other `CheckAndSetDefaults` calls if we are going to return an error anyway.

Reply:
Good thought, and a quick fix!